notebooks/hive2bq/HiveToBigquery_notebook.ipynb

```diff
@@ -288,10 +291,83 @@
   },
   {
    "cell_type": "markdown",
-   "id": "0dbaf266-91a0-46e2-8a88-b594b9301c83",
+   "id": "5cd2e0cc-62fb-4ecf-8eaf-c16565fee1cc",
    "metadata": {},
    "source": [
     "#### Step 8:\n",
+    "#### Extract Hive DDls \n",
+    "Below cell will fetch DDls of all the tables in the given database from HIVE metastore and store in the below GCP location\n",
+    "\n",
+    "gs://{TEMP_BUCKET}/hive_ddls/input/{INPUT_HIVE_DATABASE}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0c4f1a2e-e75c-4583-a22e-8972b9decaaf",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Extract HIVE DDLs from Hive Metastore\n",
+    "get_hive_ddls(INPUT_HIVE_DATABASE,TABLE_LIST,TEMP_BUCKET,spark)"
```

Comment:
Hi @NiloFreitas , Thank you for taking out time to review this. The reason I am not able to use HIVE DDL Extractor utility directly is that there are some values in the show create command which we need to avoid giving to the translation API in order to get expected result. One of the important one is location. Show create command will give location parameter but if we give the same to translation API, it will take it as an external table and will create external BQ table instead.
Example:
<img width="1366" alt="Screenshot 2023-02-05 at 4 11 50 PM" src="https://user-images.githubusercontent.com/105909213/216814348-e57d8969-94e5-4f79-9c19-0fc20605dd2f.png">

Also, if the table is created using spark, the HIVE extract utility will give error. This is the reason I have added the logic to get show create using both with and without "AS SERDE"

Reply:
What do you think of modifying the DDL extractor to get optional parameter to apply these transformations?
Like booleans: **ddl_write_disposition**, **ddl_as_internal_table**, and **consider_spark_tables**

Reply:
I am open to that but I feel that is just trying to modify a simple ddl extractor based on only our need as high chances that others may not require it. Also when I said external table I meant after the translation in the BQ side and not HIVE tables. So if I give location parameter in HIVE ddl to the translation API it will create BQ external table ddl. If we do not provide location translation API will always hive BQ internal table only no matter if the HIVE table DDL statement contains external keyword or now.

Another option is to have this TRANSLATION API as a separate utility and call the same in this notebook. Although then again we will have to create extract ddl utility on our own as we need to remove the location parameter from HIVE DDL before passing the same to translation API.

Reply:
I would suggest us to modify the DDL extractor to support these additional optional parameters, because our users will also probably use them as well, in their end-to-end use case. I think this notebook is one end-to-end more similar to real use cases, so the DDL extractor, despite its usefullness, seems to be currently not enough to attend our migration requirements. Putting all this logic in the notebook is not making the code reusable enough.

Many migrations may wish to transform Hive External Tables to Native BigQuery, so that is why I suggested the utility function to be able to "remove DDL location" while extracting them, so that the Translation API can be used in a more straightforward way.

Moreover, I think the utility templates for translation api may seem interesting but we should evaluate if spark is really needed in this step (to be treated as a dataproc template), instead of just code in the notebook as means to and end.

I would suggest to remove some of the developed logic in the notebook (IF NOT EXISTS, location, Spark+Hive tables) and refactor the DDL Extractor to accept parameters, and then have a simplified code in the notebook, keeping the translation logic in the notebook for now.

What do you think?

Cheers

Reply:
Got it, Thanks @NiloFreitas. I think those are some fair points. So to sum it up:

1. Modify DDL Extractor to accept some optional arguments. 
 ddl_write_disposition
 ddl_as_internal_table
 consider_spark_tables
 I may need your suggestions on actual use case for some of the above parameters.
 
 2. Once we are satisfied with the DDL extractor, we can move on to translation API:
I think we can have this as separate python generic script in  util folder which can be called by other templates as well. We can also have an optional argument if the user also wants to run those translated DDLs in BQ and create tables (needed for our use case). 

3. Lastly, call everything in the existing HIVE2BQ notebook.

I am going to create a new issue for DDL Extractor and get started on it.

Reply:
DDL Extractor modified and pushed to main branch. Next, will start discussion on translation API.